src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -509,6 +516,42 @@ InterpreterSelectQuery::InterpreterSelectQuery(
         query_info.additional_filter_ast = parseAdditionalFilterConditionForTable(
             settings.additional_table_filters, joined_tables.tablesWithColumns().front().table, *context);
 
+    ASTPtr parallel_replicas_custom_filter_ast = nullptr;
+    if (context->getParallelReplicasMode() == Context::ParallelReplicasMode::CUSTOM_KEY && !joined_tables.tablesWithColumns().empty())
+    {
+        if (settings.parallel_replicas_count > 1)
+        {
+            if (auto custom_key_ast = parseCustomKeyForTable(settings.parallel_replicas_custom_key, *context))
+            {
+                LOG_INFO(log, "Processing query on a replica using custom_key");
```

Comment:
Also dump key expression here?
Also, I would prefer LOG_TRACE (we had log-level devaluation).